File: 2023/go/day2/2.go
package day2

import (
	"strconv"
	"strings"

	log "github.com/sirupsen/logrus"
)

type cube struct {
	color    string
	quantity int
}

func Puzzle(input *[]byte, part2 bool) int {
	//log.SetLevel(log.DebugLevel)
	totalCubes := map[string]int{
		"red":   12,
		"green": 13,
		"blue":  14,
	}
	var possible, powers []int
	for _, line := range strings.Split(strings.TrimSpace(string(*input)), "\n") {
		content := strings.Split(strings.TrimSpace(line), ":")
		gameID, err := strconv.Atoi(strings.Split(strings.TrimSpace(content[0]), " ")[1])
		if err != nil {
			log.Error(err)
		}
		minCubes := map[string]int{
			"red":   0,
			"green": 0,
			"blue":  0,
		}
		impossible := false
		sets := strings.Split(strings.TrimSpace(content[1]), ";")
		for _, set := range sets {
			for _, spec := range strings.Split(strings.TrimSpace(set), ",") {
				value := strings.Split(strings.TrimSpace(spec), " ")
				cube := cube{
					quantity: 0,
					color:    value[1],
				}
				cube.quantity, err = strconv.Atoi(value[0])
				if err != nil {
					log.Error(err)
				}
				if !impossible && cube.quantity > totalCubes[cube.color] {
					log.Debug("Game ", gameID, " is impossible.")
					impossible = true
				}
				if minCubes[cube.color] < cube.quantity {
					minCubes[cube.color] = cube.quantity
				}
			}
		}
		if !impossible {
			possible = append(possible, gameID)
		}
		log.Debug("Game ", gameID, " minimum cubes required: red=", minCubes["red"], " green=", minCubes["green"], " blue=", minCubes["blue"])
		power := 1
		for _, v := range minCubes {
			power *= v
		}
		powers = append(powers, power)
		log.Debug("Game ", gameID, " power: ", power)
	}
	sum := 0
	for _, id := range possible {
		sum += id
	}
	if part2 {
		sum = 0
		for _, power := range powers {
			sum += power
		}
	}
	return sum
}

File: 2023/python/2.py
#!/usr/bin/env python3

import time
startTime = time.time()

def puzzle(input, part2):
    totalCubes = { "red": 12, "green": 13, "blue": 14 }
    possibleSum, powerSum = 0, 0
    for line in input.splitlines():
        content = line.split(": ")
        id = int(content[0].split(" ")[1])
        minCubes = {"red": 0, "green": 0, "blue": 0}
        impossible, power = False, 1
        sets = content[1].split("; ")
        for set in sets:
            for cube in set.split(", "):
                cube = cube.split(" ")
                if not impossible and int(cube[0]) > totalCubes[cube[1]]:
                    impossible = True
                if minCubes[cube[1]] < int(cube[0]):
                    minCubes[cube[1]] = int(cube[0])
        if not impossible:
            possibleSum += id
        for color in minCubes:
            power *= minCubes[color]
        powerSum += power
    if part2:
        return powerSum
    return possibleSum

f = open("2023/input/2", "r")
input = f.read()

print("2023 - Day 2 - Part 1:", puzzle(input, False))
print("2023 - Day 2 - Part 2:", puzzle(input, True))
print("Execution time: %.3fµs" % ((time.time() - startTime)*1000))